// @unified-latex/unified-latex-ctan/package/systeme/index.cjs
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// package/systeme/index.ts
var systeme_exports = {};
__export(systeme_exports, {
  attachSystemeSettingsAsRenderInfo: () => attachSystemeSettingsAsRenderInfo,
  environments: () => environments,
  extractVariables: () => extractVariables,
  macros: () => macros,
  parse: () => parse,
  printRaw: () => printRaw,
  systemeContentsToArray: () => systemeContentsToArray
});
module.exports = __toCommonJS(systeme_exports);

// package/systeme/provides.ts
var macros = {
  systeme: {
    signature: "s o o m",
    renderInfo: { inMathMode: true }
  },
  sysdelim: {
    signature: "m m"
  },
  syseqsep: { signature: "m" },
  sysalign: { signature: "m" },
  syssignspace: { signature: "m" },
  syseqspace: { signature: "m" },
  syslineskipcoeff: { signature: "m" },
  syseqivsign: { signature: "m" },
  sysaddeqsign: { signature: "m" },
  sysremoveeqsign: { signature: "m" },
  sysextracolonsign: { signature: "m" },
  syscodeextracol: { signature: "m" },
  sysautonum: { signature: "m" },
  syssubstitute: { signature: "m" }
};
var environments = {};

// package/systeme/libs/parser.ts
var import_unified_latex_util_match = require("@unified-latex/unified-latex-util-match");
var import_unified_latex_util_pegjs = require("@unified-latex/unified-latex-util-pegjs");
var import_unified_latex_util_pegjs2 = require("@unified-latex/unified-latex-util-pegjs");
function createMatchers({
  at = "@",
  equals = "=",
  equationSeparator = ",",
  mathOperations = ["+", "-"],
  whitelistedVariables
} = {}) {
  let isVar = (node) => import_unified_latex_util_match.match.anyString(node) && !!node.content.match(/[a-zA-Z]/);
  if (whitelistedVariables) {
    whitelistedVariables = whitelistedVariables.map((v) => import_unified_latex_util_match.match.anyString(v) ? v.content : v);
    const macros2 = whitelistedVariables.filter((v) => import_unified_latex_util_match.match.anyMacro(v));
    const strings = whitelistedVariables.filter((v) => typeof v === "string");
    const macroHash = Object.fromEntries(macros2.map((v) => [v.content, v]));
    const stringHash = Object.fromEntries(strings.map((s) => [s, s]));
    const macroMatcher = import_unified_latex_util_match.match.createMacroMatcher(macroHash);
    isVar = (node) => macroMatcher(node) || import_unified_latex_util_match.match.anyString(node) && !!stringHash[node.content];
  }
  return {
    isSep: (node) => import_unified_latex_util_match.match.string(node, equationSeparator),
    isVar,
    isOperation: (node) => mathOperations.some((op) => import_unified_latex_util_match.match.string(node, op)),
    isEquals: (node) => import_unified_latex_util_match.match.string(node, equals),
    isAt: (node) => import_unified_latex_util_match.match.string(node, at),
    isSubscript: (node) => import_unified_latex_util_match.match.macro(node, "_") && node.escapeToken === "",
    isWhitespace: import_unified_latex_util_match.match.whitespace,
    isSameLineComment: (node) => import_unified_latex_util_match.match.comment(node) && node.sameline,
    isOwnLineComment: (node) => import_unified_latex_util_match.match.comment(node) && !node.sameline
  };
}
function parse(ast, options) {
  if (!Array.isArray(ast)) {
    throw new Error("You must pass an array of nodes");
  }
  ast = (0, import_unified_latex_util_pegjs.decorateArrayForPegjs)([...ast]);
  return import_unified_latex_util_pegjs2.SystemePegParser.parse(ast, createMatchers(options || {}));
}

// package/systeme/libs/print-raw.ts
var import_unified_latex_util_print_raw = require("@unified-latex/unified-latex-util-print-raw");
function printRaw(node, root = false) {
  if (typeof node === "string") {
    return node;
  }
  if (Array.isArray(node)) {
    const sepToken = root ? " " : "";
    return node.map((tok) => printRaw(tok)).join(sepToken);
  }
  switch (node.type) {
    case "annotation":
      return `${(0, import_unified_latex_util_print_raw.printRaw)(node.marker)}${(0, import_unified_latex_util_print_raw.printRaw)(node.content)}`;
    case "item":
      return `${node.op ? (0, import_unified_latex_util_print_raw.printRaw)(node.op) : ""}${(0, import_unified_latex_util_print_raw.printRaw)(node.content)}`;
    case "equation":
      const left = node.left.map((n) => printRaw(n)).join("");
      const right = (0, import_unified_latex_util_print_raw.printRaw)(node.right);
      const equals = node.equals ? (0, import_unified_latex_util_print_raw.printRaw)(node.equals) : "";
      return `${left}${equals}${right}`;
    case "line":
      const equation = node.equation ? printRaw(node.equation) : "";
      const annotation = node.annotation ? printRaw(node.annotation) : "";
      const sep = node.sep ? (0, import_unified_latex_util_print_raw.printRaw)(node.sep) : "";
      const body = `${equation}${annotation}${sep}`;
      if (node.trailingComment) {
        return (0, import_unified_latex_util_print_raw.printRaw)([body, node.trailingComment]);
      }
      return body;
    default:
      console.warn(`Unknown node type "${node.type}" for node`, node);
      return "";
  }
}

// package/systeme/libs/systeme.ts
var import_unified_latex_util_print_raw2 = require("@unified-latex/unified-latex-util-print-raw");
var import_unified_latex_util_match2 = require("@unified-latex/unified-latex-util-match");
var import_unified_latex_util_split = require("@unified-latex/unified-latex-util-split");

// ../structured-clone/dist/index.js
(function() {
  if (typeof globalThis === "object")
    return;
  Object.defineProperty(Object.prototype, "__magic__", {
    get: function() {
      return this;
    },
    configurable: true
  });
  __magic__.globalThis = __magic__;
  delete Object.prototype.__magic__;
})();
var clone = typeof globalThis.structuredClone === "function" ? globalThis.structuredClone : (obj) => JSON.parse(JSON.stringify(obj));
function structuredClone(obj) {
  return clone(obj);
}

// package/systeme/libs/systeme.ts
var import_unified_latex_util_comments = require("@unified-latex/unified-latex-util-comments");
var import_unified_latex_util_visit = require("@unified-latex/unified-latex-util-visit");
var import_unified_latex_util_render_info = require("@unified-latex/unified-latex-util-render-info");
var import_unified_latex_util_arguments = require("@unified-latex/unified-latex-util-arguments");
var AMP = { type: "string", content: "&" };
var SEP = { type: "macro", content: "\\" };
var QUAD = { type: "macro", content: "quad" };
var PLUS = { type: "string", content: "+" };
var COLUMN_KERN_ADJUSTMENT = [
  { type: "string", content: "@" },
  {
    type: "group",
    content: [
      { type: "macro", content: "mkern" },
      { type: "string", content: "5mu" }
    ]
  }
];
function sortVariables(vars, whitelistedVariables) {
  const varMap = new Map(vars.map((v) => [v, (0, import_unified_latex_util_print_raw2.printRaw)(v)]));
  const varNames = Array.from(new Set(varMap.values()));
  varNames.sort();
  const nameToPos = whitelistedVariables ? new Map(whitelistedVariables.map((v, i) => [(0, import_unified_latex_util_print_raw2.printRaw)(v), i])) : new Map(varNames.map((name, i) => [name, i]));
  return new Map(Array.from(varMap.entries()).map(([variable, name]) => {
    return [variable, nameToPos.get(name) ?? -1];
  }));
}
function processLine(line, numVars, varOrder, hasEquals, hasAnnotation) {
  const ret = [];
  if (line.equation) {
    const nonVarItems = line.equation.left.filter((item) => item.variable == null);
    const varItems = line.equation.left.filter((item) => item.variable != null);
    let nonVarTerm = null;
    if (nonVarItems.length === 1) {
      nonVarTerm = nonVarItems[0];
    } else if (nonVarItems.length > 1) {
      nonVarTerm = __spreadProps(__spreadValues({}, nonVarItems[0]), {
        content: nonVarItems[0].content.concat(nonVarItems.slice(1).flatMap((item) => {
          if (item.op) {
            return [item.op, ...item.content];
          }
          return [PLUS, ...item.content];
        }))
      });
    }
    const allItems = nonVarTerm ? varItems.concat(nonVarTerm) : varItems;
    const indexToItem = new Map(allItems.map((item) => {
      if (item.variable == null) {
        return [numVars - 1, item];
      }
      return [varOrder.get(item.variable), item];
    }));
    let isFirstItem = true;
    for (let i = 0; i < numVars; i++) {
      const item = indexToItem.get(i);
      if (item) {
        if (isFirstItem && (import_unified_latex_util_match2.match.string(item.op, "+") || item.op == null)) {
          ret.push([]);
          ret.push(item.content);
        } else {
          ret.push([item.op || PLUS]);
          ret.push(item.content);
        }
        isFirstItem = false;
      } else {
        ret.push([]);
        ret.push([]);
      }
    }
    if (hasEquals) {
      const equalsPart = (line.equation.equals ? [line.equation.equals] : []).concat(line.equation.right);
      ret.push(equalsPart);
    }
  }
  if (hasAnnotation) {
    ret.push(line.annotation ? line.annotation.content : []);
  }
  return ret;
}
function arraySpecToSpacedArraySpec(spec, hasAnnotation) {
  const annotationSpec = hasAnnotation ? spec.charAt(spec.length - 1) : "";
  const bodySpec = hasAnnotation ? spec.slice(0, spec.length - 1) : spec;
  const bodyStrings = Array.from(bodySpec).map((x) => [
    { type: "string", content: x }
  ]);
  const body = (0, import_unified_latex_util_split.arrayJoin)(bodyStrings, COLUMN_KERN_ADJUSTMENT);
  return annotationSpec ? body.concat({ type: "string", content: annotationSpec }) : body;
}
function extractVariables(nodes) {
  return nodes.flatMap((node) => {
    if (node.type === "line" && node.equation) {
      return extractVariables(node.equation.left);
    }
    if (node.type === "equation") {
      return node.left.flatMap((item) => item.variable ? [item.variable] : []);
    }
    if (node.type === "item") {
      return node.variable ? [node.variable] : [];
    }
    return [];
  });
}
function normalizeVariableWhitelist(vars) {
  if (!vars) {
    return null;
  }
  const normalized = vars.map((v) => typeof v === "string" ? { type: "string", content: v } : v);
  const ret = normalized.filter((v) => (import_unified_latex_util_match2.match.anyMacro(v) || import_unified_latex_util_match2.match.anyString(v)) && !import_unified_latex_util_match2.match.string(v, " ") && !import_unified_latex_util_match2.match.whitespace(v));
  return ret;
}
function systemeContentsToArray(nodes, options) {
  nodes = structuredClone(nodes);
  (0, import_unified_latex_util_comments.deleteComments)(nodes);
  const { properSpacing = true, whitelistedVariables } = options || {};
  const coercedWhitelistedVariables = normalizeVariableWhitelist(whitelistedVariables);
  const systemeAst = parse(nodes, { whitelistedVariables });
  const vars = extractVariables(systemeAst);
  const varOrder = sortVariables(vars, coercedWhitelistedVariables);
  let numVars = coercedWhitelistedVariables ? coercedWhitelistedVariables.length : Math.max(...Array.from(varOrder.values())) + 1;
  if (systemeAst.some((line) => {
    if (line.equation) {
      return line.equation.left.some((item) => item.variable == null);
    }
  })) {
    numVars += 1;
  }
  const hasEquals = systemeAst.some((line) => line.equation && line.equation.equals);
  const hasAnnotation = systemeAst.some((line) => line.annotation);
  let rows = systemeAst.map((line) => processLine(line, numVars, varOrder, hasEquals, hasAnnotation));
  const noLeadingOperation = rows.every((row) => row[0].length === 0);
  let arraySignature = Array.from({ length: numVars }, () => "cr").join("");
  if (noLeadingOperation) {
    arraySignature = arraySignature.slice(1);
    rows = rows.map((row) => row.slice(1));
  }
  if (hasEquals) {
    arraySignature += "l";
  }
  if (hasAnnotation) {
    arraySignature += "l";
    rows = rows.map((row) => {
      if (row[row.length - 1].length === 0) {
        return row;
      }
      return [
        ...row.slice(0, row.length - 1),
        [QUAD, { type: "whitespace" }, ...row[row.length - 1]]
      ];
    });
  }
  const arraySignatureWithSpacing = properSpacing ? arraySpecToSpacedArraySpec(arraySignature, hasAnnotation) : [{ type: "string", content: arraySignature }];
  const bodyRows = rows.map((row) => (0, import_unified_latex_util_split.arrayJoin)(row, AMP));
  const body = (0, import_unified_latex_util_split.arrayJoin)(bodyRows, SEP);
  const ret = {
    type: "environment",
    env: "array",
    args: [
      {
        type: "argument",
        openMark: "{",
        closeMark: "}",
        content: arraySignatureWithSpacing
      }
    ],
    content: body
  };
  return ret;
}
function attachSystemeSettingsAsRenderInfo(ast) {
  const systemeMatcher = import_unified_latex_util_match2.match.createMacroMatcher(["systeme", "sysdelim"]);
  (0, import_unified_latex_util_visit.visit)(ast, (nodes, info) => {
    if (!info.context.inMathMode || !nodes.some(systemeMatcher)) {
      return;
    }
    const systemeLocations = nodes.flatMap((node, i) => import_unified_latex_util_match2.match.macro(node, "systeme") ? i : []);
    const sysdelimLocations = nodes.flatMap((node, i) => import_unified_latex_util_match2.match.macro(node, "sysdelim") ? i : []);
    if (systemeLocations.length === 0 || sysdelimLocations.length === 0) {
      return;
    }
    for (const i of systemeLocations) {
      const lastSysdelim = Math.max(...sysdelimLocations.filter((loc) => loc < i));
      if (lastSysdelim >= 0) {
        const node = nodes[i];
        const sysdelimMacro = nodes[lastSysdelim];
        if (!import_unified_latex_util_match2.match.anyMacro(sysdelimMacro)) {
          throw new Error(`Expecting sysdelim macro but found "${(0, import_unified_latex_util_print_raw2.printRaw)(sysdelimMacro)}"`);
        }
        const args = (0, import_unified_latex_util_arguments.getArgsContent)(sysdelimMacro);
        (0, import_unified_latex_util_render_info.updateRenderInfo)(node, { sysdelims: args });
      }
    }
  }, {
    test: Array.isArray,
    includeArrays: true
  });
}
